Doc/reference/datamodel.rst

```diff
@@ -970,10 +970,20 @@ Internal types
 
       .. index::
          single: f_trace (frame attribute)
+         single: f_trace_lines (frame attribute)
+         single: f_trace_opcodes (frame attribute)
          single: f_lineno (frame attribute)
 
       Special writable attributes: :attr:`f_trace`, if not ``None``, is a function
-      called at the start of each source code line (this is used by the debugger);
+      called for various events during code execution (this is used by the debugger).
+      Normally an event is triggered for each new source line - this can be
+      disabled by setting :attr:`f_trace_lines` to :const:`False`.
+
+      Implementations *may* allow per-opcode events to be requested by setting
+      :attr:`f_trace_opcodes` to :const:`True`. This is intended specifically
+      for interpreter signal safety testing, and may have surprising results
```

Comment:
perhaps change the wording to:

_and may result in surprising undefined behavior if an exception escapes the trace function._

i'm trying to make it more explicit that we do not guarantee any behavior when exceptions are allowed to escape during opcode tracing.  changing "are raised by" to something involving "escape" makes it clear that exceptions can be raised and handled.  the important thing is that they are resolved before the function exits.